### PR TITLE
CSCQVAIN-141: Keep front-end and back-end session status better in sync

### PR DESCRIPTION
**Pull request by tahme**
Check if the session still exists in the back-end when resuming a session. Logout from back-end when logout is clicked.

### CODE REVIEW

File: src/auth/plugin.js [hunk 1]
```diff
@@ -6,10 +6,10 @@ function addGlobalGuard(router, auth, loginPage) {
 		loginPage = "/login"
 	}
 
-	router.beforeEach((to, from, next) => {
-		// check for login token in local storage
+	router.beforeEach(async (to, from, next) => {
+		// check for existing session
 		if (!auth.loggedIn) {
-			auth.localLogin()
+			await auth.resumeSession()
```
Comment (tahme) [inline src/auth/plugin.js]: On initial page load for a potentially-logged-in user (i.e. we might have an existing session but need to poll the back-end to find out for sure), this causes a tiny moment where the user is shown as not logged-in and router-view is empty, which annoys me slightly.

Comment (jppiiroinen) [reply src/auth/plugin.js]: there should be a loading indicator, until the session information is known

Comment (tahme) [reply src/auth/plugin.js]: I added a loading indicator to the user dropdown button. 

I also tried using a loading indicator for the page content until the session is loaded, but I wasn't sure it was better than without the indicator (especially in the datasets page that liked to freeze animations until the dataset table is finished), so I left it out for now.

File: src/auth/auth.js [hunk 2]
```diff
@@ -1,4 +1,6 @@
+import axios from 'axios'
 import {parseJwt, getRandomString} from './jwt.js'
+import Vue from 'vue'
```
Comment (jppiiroinen) [inline src/auth/auth.js:4]: related to auth.js, could you check if that getId function needs to be fixed, as that is missing return.

Comment (tahme) [reply src/auth/auth.js:4]: Good catch. It was broken in multiple ways and not used anywhere so I removed it now.

File: src/auth/auth.js [hunk 3]
```diff
@@ -67,8 +65,11 @@ function filterGroups(prefix, groups) {
 	return groups.filter(grp => grp.startsWith(prefix)).map(grp => grp.substring(prefix.length)).filter(grp => !isNaN(grp))
 }
 
-function Auth(url) {
-	this.url = url
+function Auth(loginUrl, logoutUrl, sessionsUrl) {
+	// reactive indicator for when session is being loaded from the back-end
+	this.loading = Vue.observable({
```
Comment (jppiiroinen) [inline src/auth/auth.js:70]: I was trying to understand the memory handling on this, as this adds a listener to the object, but how do we actually remove the listener from this.loading? that `__ob__` thing? and wondering does this cause a memory leak, the `this.loading` does not seem to have any `destroy` functions or anything which could be used to remove the listener.

Comment (tahme) [reply src/auth/auth.js:70]: I don't think you generally remove the `__ob__` thing aka Observer (implementation [here](https://github.com/vuejs/vue/blob/dev/src/core/observer/index.js)). It should be garbage collected together with the entire `this.loading` object when there are no references / reactive dependencies to it.

I guess that if you no longer need the object to be reactive, you could convert it back to a plain javascript object with something like `this.loading = JSON.parse(JSON.stringify(this.loading))`, but that won't help with existing watchers that depend on it.

Comment (jppiiroinen) [reply src/auth/auth.js:70]: I wasn't able to find anything related to this either, lets go with this now and fix it later if this will cause issues.

**Review by jppiiroinen — APPROVED** [review]

**Pull request merged**